cdap-watchdog/src/main/java/co/cask/cdap/logging/write/LogCleanup.java

```diff
@@ -109,13 +127,191 @@ public Void call() throws Exception {
+                             final SetMultimap<String, Location> parentDirs) throws Exception {
+    LOG.info("Starting deletion of log files older than {} without metadata", tillTime);
+    List<NamespaceMeta> namespaces = namespaceQueryAdmin.list();
+    // For all the namespaces present in NamespaceMeta, gather log files from disk and log meta. Then delete the log
+    // files for which metadata is not present
+    for (NamespaceMeta namespaceMeta : namespaces) {
+      final NamespaceId namespaceId = namespaceMeta.getNamespaceId();
+      final Location namespacedLogBaseDir = LoggingContextHelper.getNamespacedBaseDirLocation(namespacedLocationFactory,
+                                                                                              logBaseDir, namespaceId,
+                                                                                              impersonator);
+      // processor to get list of log files older than retention duration
+      final Processor<LocationStatus, Set<Location>> diskFilesProcessor = getDiskFilesProcessor(namespaceId, tillTime);
+      if (namespacedLogBaseDir.exists()) {
+        // Get all the disk log files on disk for a given namespace
+        Set<Location> diskFileLocations = getDiskLocations(namespaceId, namespacedLogBaseDir, diskFilesProcessor);
+        // remove log files from diskFileLocations for which metadata is present.
+        removeLocationsWithMeta(tillTime, namespaceId, diskFileLocations);
+
+        // delete all the disk locations for which metadata is not present
+        for (final Location locationToDelete : diskFileLocations) {
+          try {
+            deleteLogFiles(parentDirs, namespaceId, namespacedLogBaseDir.toString(), locationToDelete);
+          } catch (Exception e) {
+            LOG.warn("Got exception when deleting path {}", locationToDelete, e);
+          }
+          namespacedLogBaseDirMap.put(namespacedLogBaseDir.toString(), namespaceId);
+        }
+      }
+    }
+  }
+
+  /**
+   * Remove all the disk files from diskFileLocations for which meta data is present
+   * @param tillTime time till the meta data will be deleted.
+   * @param namespaceId namespace for which metadata needs to be scanned
+   * @param diskFileLocations log files present on disk for given namespace
+   */
+  private void removeLocationsWithMeta(final long tillTime, NamespaceId namespaceId,
+                                       final Set<Location> diskFileLocations) {
+    // get the logging context for a given namespace
+    LoggingContext loggingContext = new LogNamespaceLoggingContext(namespaceId.getNamespace());
+    FileMetaDataManager.TableKey nextTableKey = new FileMetaDataManager.TableKey(loggingContext.getLogPartition(),
+                                                                                 null);
+    Processor<URI, Set<URI>> metaProcessor = getMetaFilesProcessor();
```

Comment:
`metaProcessor` will need to be recreated or rest every time in the do-while loop, right? Otherwise it will collect all the files cumulatively in every loop, right?